APMrover2/mode.cpp

```diff
@@ -145,6 +145,26 @@ void Mode::get_pilot_desired_lateral(float &lateral_out)
     lateral_out = rover.channel_lateral->get_control_in();
 }
 
+// decode pilot's input and return heading_out (in cd) and speed_out (in m/s)
+void Mode::get_pilot_desired_heading_and_speed(float &heading_out, float &speed_out)
+{
+    // get steering and throttle in the -1 to +1 range
+    float desired_steering = constrain_float(rover.channel_steer->get_control_in() / 4500.0f, -1.0, 1.0f);
+    float desired_throttle = constrain_float(rover.channel_throttle->get_control_in() / 100.0f, -1.0f, 1.0f);
+
+    // calculate angle of input stick vector
+    heading_out = wrap_360_cd(atan2f(desired_steering, desired_throttle) * DEGX100);
+
+    // calculate magnitude of input stick vector
+    float magnitude_max = 1.0f;
+    float magnitude = safe_sqrt(sq(desired_throttle) + sq(desired_steering));
```

Comment:
Could notionally avoid this safe_sqrt in the (rare?) case that magnitude^2 > magnitude_max^2

Reply:
Possibly clearer version (depends on point of view), would be
```
float magnitude = min(safe_sqrt(sq(desired_throttle) + sq(desired_steering)), 1.0);
```

Which has the advantage of not needing the explicit if/const.

Reply:
great, thanks.. I've greatly shortened the calcs for this section..